src/rfcs/0003-x86_64-system-abi.md

```diff
@@ -0,0 +1,116 @@
+
+## Motivation
+
+Toolchains compiling to Lilium on x86_64, as well as hand-written assembly, need to be able to produce code that can interact with system libraries as well as system functions, using standardized conventions.
+
+## Informative Explanation
+
+The x86_64 Architecture is one of the primary architectures targetted by the Lilium Operating System. To esnure compatibility of compiled programs, we define an set of conventions for the ABI, 
+The conventions consists of the Calling Convention for both Userspace and System Calls from the Kernel, as well as layouts and representations of language types and certain vocabulary types in the standard library,
+
+## Normative Text
+
+### Userspace ABI
+
+Except as specified below, all system libraries in userspace obey the [x86-64 psABI]. 
+
+The C `main` function, and function pointers passed to system libraries, is expected to obey this ABI, and may have the following signatures:
+
+* `int main(void)`
+* `int main(int argc, char** argv)` or `int main(long argc, char** argv)`
+* `int main(int argc, char** argv, char** envp)` or `int main(long argc, char** argv, char** envp)`
+
+Regardless of whether `argc` is defined as `int` or `long`, it will contain the same value, unless the value cannot be represented as `int`. 
+
+(Note: Signatures using `long argc` are non-standard and are not portable)
+
+(Note: It is recommended, but not strictly required, that all userspace code obey this ABI).
+
+### System Call ABI
+
+The System Call ABI uses a modify form of the calling convention from the [x86-64 psABI]. It uses the same type layouts as the Userspace ABI.
+
+System Calls (defined by the kernel) are invoked by the `syscall` instruction. The system function number 
+
+The following changes apply:
+* On entry, `eax` contains the system function number. The top 32-bits of `rax` must be clear.
+* On exit:
+    * If the function returns `void`, both `rax` and `rdx` are undefined
+    * If the function returns `SysResult`, `rax` contains the return value - the negative error number on error, positive or `0` if succesful, `rdx` is undefined.
+    * If the function returns `SysResult2<T>` (`T` must be up to 8 bytes in size, and have class INTEGER), `rax` contains the error or `0` if successful, and `rdx` contains the value if successful (`rdx` is undefined if `rax` contains an error)
+    * If the function returns any other type, that type must be at most one eightbyte and have class INTEGER. The value is in `rax` and `rdx` is undefined.
+* There may be at most 6 eightbytes of parameters, each of either class MEMORY or INTEGER
+* The fourth INTEGER eightbyte for parameters (including pointers for MEMORY types) is passed in `r10`, not in `rcx`. If the system function uses fewer than 4 eightbytes, `r10` is not used for the `syscall` (caller saved/volatile).
```

Comment:
Why is this? Seems interesting and arbitrary; may be worth a rationale note

Reply:
I see that it's a quirk of the `syscall` instruction. Could still be nice to mention specifically.